raiden-ts/src/utils/ethers.ts

```diff
@@ -78,54 +82,78 @@ export function fromEthersEvent<T>(
-  fromBlock?: number,
+  opts?: {
+    fromBlock?: number;
+    confirmations?: number | Observable<number>;
+    blockNumber$?: Observable<number>;
+    onPastCompleted?: (elapsedMs: number) => void;
+  },
 ): Observable<T>;
 /**
  * Like rxjs' fromEvent, but event can be an EventFilter
  *
  * @param target - Object to hook event listener, maybe a Provider or Contract
  * @param event - EventFilter or string representing the event to listen to
- * @param confirmations - After how many blocks a tx is considered confirmed
- * @param fromBlock - Block since when to fetch events from
+ * @param opts - Options object
+ * @param opts.fromBlock - Block since when to fetch events from
+ * @param opts.confirmations - After how many blocks a tx is considered confirmed; if observable,
+ *    it should have a value at subscription time, like a ReplaySubject(1);
+ * @param opts.blockNumber$ - New blockNumber observable
+ * @param opts.onPastCompleted - Callback when first/past blocks scan completes
  * @returns Observable of target.on(event) events
  */
 export function fromEthersEvent<T>(
   target: JsonRpcProvider,
   event: EventType,
-  confirmations = 5,
-  fromBlock?: number,
+  {
+    fromBlock,
+    confirmations,
+    blockNumber$,
+    onPastCompleted,
+  }: {
+    fromBlock?: number;
+    confirmations?: number | Observable<number>;
+    blockNumber$?: Observable<number>;
+    onPastCompleted?: (elapsedMs: number) => void;
+  } = {},
 ) {
   if (typeof event === 'string' || Array.isArray(event))
     return fromEventPattern<T>(
       (handler: Listener) => target.on(event, handler),
       (handler: Listener) => target.removeListener(event, handler),
     ) as Observable<T>;
 
-  const range = confirmations * 2; // half for confirmed, half for unconfirmed logs
-  const blockQueue: number[] = []; // sorted 'fromBlock' queue, at most of [range] size
-  return defer(() => {
-    if (!fromBlock) {
-      // 'resetEventsBlock' is private, set at [[Raiden]] constructor, so we need 'any'
-      let resetBlock: number = (target as any)._lastBlockNumber;
-      const innerBlockNumber = target.blockNumber;
-      resetBlock =
-        resetBlock && resetBlock > 0
-          ? resetBlock
-          : innerBlockNumber && innerBlockNumber > 0
-          ? innerBlockNumber
-          : 1;
-      fromBlock = resetBlock - confirmations;
-    }
-    // starts 'blockQueue' with subscription-time's resetEventsBlock
-    blockQueue.splice(0, blockQueue.length, fromBlock);
+  const confirmations$ = !confirmations
+    ? of(5)
+    : typeof confirmations === 'number'
+    ? of(confirmations)
+    : confirmations;
+  const blockQueue: number[] = []; // sorted 'fromBlock' queue, at most of [confirmations * 2] size
+  let start = Date.now();
+  return confirmations$.pipe(
+    distinctUntilChanged(),
+    mergeWith((confirmations) => {
+      if (!fromBlock) {
+        // 'resetEventsBlock' is private, set at [[Raiden]] constructor, so we need 'any'
```

Comment:
`resetEventsBlock` sets `_lastBlockNumber`, and since that's the public interface, I chose to reference it here, but this comment is removed on current HEAD, and instead I did it by code, using the` resetBlock` var name.